[Quant][Inductor] Enable lowering of qcat

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107951
* #106958
* __->__ #106838
* #106836

**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

leslie-fang-intel committed:
Enable lowering of qcat

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jgong5 @eellison Could you help to take a look of this PR?

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -442,10 +442,55 @@ def _register_quantization_maxpool2d():
+def _register_quantization_cat():
+    dequantize_cat_pattern = CallFunction(
+        aten.cat.default,
+        ListOf(_raw_dequantize_per_tensor_activation_pattern),
```
The rewrite is only valid when the dequantization parameters are the same for all the input tensors to concat? Should we check it?

leslie-fang-intel replied:
Thanks for the comment. Yean, we need to check the quantization parameters and add it. Please help to take a look again.

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -442,10 +442,55 @@ def _register_quantization_maxpool2d():
+    def qcat(match: Match, inputs, dim, **kwargs):
+        # Refer to the implementation
+        # https://github.com/pytorch/pytorch/blob/416bf4e3e7c5f90bbc306e55501881b4f8bde981/torch/_inductor/fx_passes/post_grad.py#L234
+        # the bias, mat1, mat2 are kept order in the inputs.
```
Thanks and modify the comments here.

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -442,17 +443,48 @@ def _register_quantization_maxpool2d():
+            )
+            for mul_node in mul_nodes
+        ]
+        if not all(math.isclose(scale, scales[0], rel_tol=0.05) for scale in scales):
```
If the qparams are constructed from shared params config, the scales/zps should exactly match? Why using such a big tolerance here?

leslie-fang-intel replied:
Yes, I think so. Change the tolerance to 1e-5.

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -928,6 +928,15 @@ def as_strided_copy(x, size, stride, storage_offset=None):
 
 @register_lowering(aten.cat)
 def cat(inputs, dim=0):
+    if all(input.get_dtype() is torch.uint8 for input in inputs):
```
would it make more sense to register a separate, quantized cat op? Is this different than `quantized::cat` ?

leslie-fang-intel replied:
yean, I think `aten::cat` already cover this functionality and make the least changes. Since `quantized::cat` assume inputs as `quantized tensor`, we need to make extra change to the implementation of `quantized::cat` to make it support of `plain tensor` with `uint8` data type.

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable lowering of qcat"


**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
